tachyon/crypto/hashes/sponge/poseidon2/poseidon2_params.h

```diff
@@ -0,0 +1,109 @@
+#ifndef TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON2_POSEIDON2_PARAMS_H_
+#define TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON2_POSEIDON2_PARAMS_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <type_traits>
+
+#include "tachyon/base/bits.h"
+#include "tachyon/math/elliptic_curves/bn/bn254/fr.h"
+#include "tachyon/math/finite_fields/finite_field_traits.h"
+
+namespace tachyon::crypto {
+
+template <typename F, size_t Width, uint64_t Alpha>
+constexpr size_t GetPoseidon2PartialRounds() {
+  using PrimeField = math::MaybeUnpack<F>;
+
+  if constexpr (PrimeField::Config::kModulusBits == 31) {
+    if constexpr (Width == 16) {
+      switch (Alpha) {
+        case 3:
+          return 20;
+        case 5:
+          return 14;
+        case 7:
+          return 13;
+        case 9:
+          return 13;
+        case 11:
+          return 13;
+      }
+    } else if constexpr (Width == 24) {
+      switch (Alpha) {
+        case 3:
+          return 23;
+        case 5:
+          return 22;
+        case 7:
+          return 21;
+        case 9:
+          return 21;
+        case 11:
+          return 21;
+      }
+    }
+  } else if constexpr (PrimeField::Config::kModulusBits == 64) {
+    if constexpr (Width == 8) {
+      switch (Alpha) {
+        case 3:
+          return 41;
+        case 5:
+          return 27;
+        case 7:
+          return 22;
+        case 9:
+          return 19;
+        case 11:
+          return 17;
+      }
+    } else if constexpr (Width == 16) {
+      switch (Alpha) {
+        case 3:
+          return 42;
+        case 5:
+          return 27;
+        case 7:
+          return 22;
+        case 9:
+          return 20;
+        case 11:
+          return 18;
+      }
+    } else if constexpr (Width == 24) {
+      switch (Alpha) {
+        case 3:
+          return 47;
+        case 5:
+          return 27;
+        case 7:
+          return 22;
+        case 9:
+          return 20;
+        case 11:
+          return 18;
+      }
+    }
+  }
+  return 56;
```

Comment:
How about this code? Also, I think it would be a good idea to add `NOTREACHED()` where appropriate to the parts where it is needed!

```suggestion
  constexpr auto SelectRounds = [](uint64_t alpha,
                                   const std::array<size_t, 5>& rounds) {
    switch (alpha) {
      case 3:
        return rounds[0];
      case 5:
        return rounds[1];
      case 7:
        return rounds[2];
      case 9:
        return rounds[3];
      case 11:
        return rounds[4];
      default:
        return size_t{56};
    }
  };

  if constexpr (PrimeField::Config::kModulusBits == 31) {
    if constexpr (Width == 16) {
      return SelectRounds(Alpha, std::array<size_t, 5>{20, 14, 13, 13, 13});
    } else if constexpr (Width == 24) {
      return SelectRounds(Alpha, std::array<size_t, 5>{23, 22, 21, 21, 21});
    }
  } else if constexpr (PrimeField::Config::kModulusBits == 64) {
    if constexpr (Width == 8) {
      return SelectRounds(Alpha, std::array<size_t, 5>{41, 27, 22, 19, 17});
    } else if constexpr (Width == 16) {
      return SelectRounds(Alpha, std::array<size_t, 5>{42, 27, 22, 20, 18});
    } else if constexpr (Width == 24) {
      return SelectRounds(Alpha, std::array<size_t, 5>{47, 27, 22, 20, 18});
    }
  }
  return size_t{56};
```